sled-agent/src/hardware/illumos/disk.rs

```diff
@@ -0,0 +1,407 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! illumos-specific mechanisms for parsing disk info.
+
+use crate::hardware::illumos::gpt;
+use crate::hardware::{DiskError, DiskPaths, DiskVariant, Partition};
+use crate::illumos::zpool::ZpoolName;
+use slog::Logger;
+use std::path::PathBuf;
+use uuid::Uuid;
+
+#[cfg(test)]
+use crate::illumos::zpool::MockZpool as Zpool;
+#[cfg(not(test))]
+use crate::illumos::zpool::Zpool;
+
+// The expected layout of an M.2 device within the Oxide rack.
+//
+// Partitions beyond this "expected partition" array are ignored.
+const M2_EXPECTED_PARTITION_COUNT: usize = 6;
+static M2_EXPECTED_PARTITIONS: [Partition; M2_EXPECTED_PARTITION_COUNT] = [
+    Partition::BootImage,
+    Partition::Reserved,
+    Partition::Reserved,
+    Partition::Reserved,
+    Partition::DumpDevice,
+    Partition::ZfsPool,
+];
+
+// The expected layout of a U.2 device within the Oxide rack.
+//
+// Partitions beyond this "expected partition" array are ignored.
+const U2_EXPECTED_PARTITION_COUNT: usize = 1;
+static U2_EXPECTED_PARTITIONS: [Partition; U2_EXPECTED_PARTITION_COUNT] =
+    [Partition::ZfsPool];
+
+fn parse_partition_types<const N: usize>(
+    path: &PathBuf,
+    partitions: &Vec<impl gpt::LibEFIPartition>,
+    expected_partitions: &[Partition; N],
+) -> Result<Vec<Partition>, DiskError> {
+    if partitions.len() != N {
+        return Err(DiskError::BadPartitionLayout {
+            path: path.clone(),
+            why: format!(
+                "Expected {} partitions, only saw {}",
+                partitions.len(),
+                N
+            ),
+        });
+    }
+    for i in 0..N {
+        if partitions[i].index() != i {
+            return Err(DiskError::BadPartitionLayout {
+                path: path.clone(),
+                why: format!(
+                    "The {i}-th partition has index {}",
+                    partitions[i].index()
+                ),
+            });
+        }
+
+        // NOTE: If we wanted to, we could validate additional information about
+        // the size, GUID, or name of the partition. At the moment, however,
+        // we're relying on the index within the partition table to indicate the
+        // "intent" of the partition.
+    }
+
+    Ok(expected_partitions.iter().map(|p| p.clone()).collect())
```

Comment:
I think this could be
```suggestion
    Ok(expected_partitions.to_vec())
```
although that might need a `Copy` on `Partition`, not sure.

Reply:
Updated, thanks!